core/src/View/Hooks/Kokkos_ViewHooks.hpp

```diff
@@ -0,0 +1,147 @@
+#ifndef KOKKOS_EXPERIMENTAL_ACCESSORHOOKS_HPP
+#define KOKKOS_EXPERIMENTAL_ACCESSORHOOKS_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <functional>
+#include <mutex>
+
+namespace Kokkos {
+namespace Experimental {
+
+class ViewHolder;
+class ConstViewHolder;
+
+namespace Detail {
+template <typename View>
+using copy_subscription_function_type = void (*)(View &, const View &);
+
+template <template <typename> class Invoker, typename... Subscribers>
+struct invoke_subscriber_impl;
+
+template <template <typename> class Invoker>
+struct invoke_subscriber_impl<Invoker> {
+  template <typename... Args>
+  static void invoke(Args &&...) {}
+};
+
+template <template <typename> class Invoker, typename Subscriber,
+          typename... RemSubscribers>
+struct invoke_subscriber_impl<Invoker, Subscriber, RemSubscribers...> {
+  template <typename... Args>
+  static void invoke(Args &&... args) {
+    Invoker<Subscriber>::call(std::forward<Args>(args)...);
+    invoke_subscriber_impl<Invoker, RemSubscribers...>::invoke(
+        std::forward<Args>(args)...);
```

Comment:
I think you're right, I'll see if there is a way around this. I think an object could be moved twice here

Reply:
You can't forward twice, because if it is an r-value reference, it ends up being moved from twice.

You can, however, do a `forward` for only the last use, as in:

```
Invoker<Subscriber>::call(args...);
invoke_subscriber_impl<Invoker, RemSubscribers...>::invoke(std::forward<Args>(args)...);

```

In `call(args...)`, `args` are all l-value references because they have a name, so this should just work.

Reply:
Since the subscribers only take view paramaters for now, I just templated on that and used l-value references instead of forwarding. If we want to hook into arbitrary constructors in the future we would have to change this back to a forwarding scheme but I don't think it's important right now.